### PR TITLE
Fix snapshot automounting in mount namespaces/bind mount/...

### PR DESCRIPTION
**Pull request by felixdoerre**
### Motivation and Context
The auto-mounts for snapshots in the `.zfs` directory working on linux are a pretty big hack, even invoking the `mount` executable from kernel. This required follow up hacks to fix up problems that resulted from this handling including problems occurring if the visible location of a zfs mountpoint was moved using `pivot_root`. Because the `mount` executable is always executed in the primary `init`'s mount-ns/chroot snapshot automounts currently doesn't work at all in mount namespaces or on bind-mounts.

### State

The new testcases show the provided improved functionality and additionally document the points where the new behavior still might leave things to be desired. The new code still requires some cleanup (e.g. removing/cleaning up debug output, removing code hidden with the `USE_FC_MOUNT` macro, removing the `USE_FC_MOUNT` macro).

I'm opening this pull request now, to get feedback if the functionality as demonstrated by the new test cases is good/consistent enough for this change to get merged as an intermediate stage, leaving debugging/fixing the remaining corner cases for a future pull request. This would help me figure out if cleaning up the code at the current stage is worth the effort or if I need to continue debugging further first.

### Description
Several new mount-testcases testing the auto-mounting/unmounting of snapshots in various scenarios: `'mount_chroot', 'mount_bind', 'mount_ns', 'mount_ns_pivot', 'mount_autoshrink'`
An addition to the SPL allowing to register mount-points for auto-expiry after a specified time.
Changes to the auto-mount code to not use the mount binary anymore but directly automount the snapshots.

### How Has This Been Tested?
The new mount testcases run through with the code as it is now.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/os/linux/zfs/zfs_ctldir.c [hunk 1]
```diff
@@ -1102,6 +1137,53 @@ zfsctl_snapshot_mount(struct path *path, int flags)
...
+	if (!mnt || IS_ERR(mnt)) {
+		error = PTR_ERR(mnt);
+		goto error;
+	} else {
+		zfs_dbgmsg("yielding vfsmount %p", mnt);
+		mnt->mnt_flags |= MNT_SHRINKABLE;
+		mntget(mnt);
+		printk("adding mount to expiry list with timeout %d.",
+		    zfs_expire_snapshot);
+		spl_add_mount_to_expire(mnt, zfs_expire_snapshot);
```
Comment (youzhongyang) [inline module/os/linux/zfs/zfs_ctldir.c:1172]: zfs_expire_snapshot <= 0 means 'do not expire' but here passing 0 to spl_add_mount_to_expire() seems to expire the mount anyway. It is very important to never expire a snapshot mount when used over nfs.

Comment (felixdoerre) [reply module/os/linux/zfs/zfs_ctldir.c:1172]: Interesting, I was not aware of this semantics of `zfs_expire_snapshot`, I think we should add a test case and preserve that semantics.

However I don't understand how this is related to usage over nfs. Shouldn't the snapshots just get automounted/autounmounted over nfs like they do regularly?

Comment (youzhongyang) [reply module/os/linux/zfs/zfs_ctldir.c:1172]: Once the snapshot is accessed over nfs, if it is auto-unmounted, the nfs client will get 'stale file handle' and other strange error when trying to access the snapshot again.

Comment (felixdoerre) [reply module/os/linux/zfs/zfs_ctldir.c:1172]: So currently if you want to expose auto-mounted snapshots via nfs you always configure `zfs_expire_snapshot = 0`? Maybe that behavior needs separate investigation, as I would consider that to be a bug as well.

Comment (youzhongyang) [reply module/os/linux/zfs/zfs_ctldir.c:1172]: Correct. That's how we ended up doing. I wouldn't say it is a bug, what can you expect when a cross mount (snapshot) is gone (unmounted) from the server?

Comment (felixdoerre) [reply module/os/linux/zfs/zfs_ctldir.c:1172]: I would expect it being re-mounted when it is being accessed. Actually the nfs mount (i.e. the nfs client) should also expire after not being access for some time. I've experimented with the nfs mount and observed the following behavior:
State: ds mounted, snap not mounted, nfs mounted, nfs-snap not moutned
Action: ls on nfs-snap, success
State: ds mounted, snap mounted, nfs mounted, nfs-snap mounted
Action: umount snap, fails with "busy"
State: ds mounted, snap mounted, nfs mounted, nfs-snap mounted
Action: Wait for more than 10mins
State: ds mounted, snap not mounted, nfs mounted, nfs-snap not mounted
Action: ls on nfs-snap, failure (stale file handle)
I would have expected: success and state: ds mounted, snap mounted, nfs mounted, nfs-snap mounted.

I would (naively) expect that you cannot unmount the snap while a client still has the snapshot via nfs automounted and once all clients have expired the server can expire as well.
